src/sumcheck/symbolic.rs

```diff
@@ -0,0 +1,217 @@
+use crate::{fields::FieldElement, sumcheck::constraints::LinearConstraintLhsTerm};
+use std::ops::{Add, AddAssign, Mul, MulAssign};
+
+/// A symbolic expression, used to accumulate symbolic terms that contribute to a circuit layer's
+/// linear constraint.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct SymbolicExpression<FieldElement> {
+    constraint_number: usize,
+    terms: Vec<SymbolicTerm<FieldElement>>,
+}
```

Comment:
FYI, a potential footgun: the derived equality comparisons would say that two `SymbolicExpression` values are different, even if the mathematical expressions they represent are equal, if they distribute the constant term differently between the `known` fields of different `SymbolicTerm`s. I assume we're only using the `PartialEq` implementation in tests, and they're passing, so this may not matter in practice. (Putting my optimization hat on, having only one `FieldElement` for the constant term in an expression would save memory)